File: TypeScript/algorithms/sorting/quick-sort.ts
import {
  swap,
  CompareFn,
  strAscendingOrderFn,
  strDescendingOrderFn
} from '../../util'

function partition(
  array: unknown[],
  left: number,
  right: number,
  compareFn: CompareFn) {

  const pivot = array[Math.floor((right + left) / 2)]
  let i = left
  let j = right
  const srtcompareFn = compareFn.toString()

  // console.log('pivot is ' + pivot + '; left is ' + left + '; right is ' + right)
  if (srtcompareFn === strAscendingOrderFn) {
    while (i <= j) {
      while (array[i] < pivot) {
        i++
        // console.log('i = ' + i)
      }

      while (array[j] > pivot) {
        j--
        // console.log('j = ' + j)
      }

      if (i <= j) {
        // console.log('swap ' + array[i] + ' with ' + array[j])
        swap(array, i, j)
        i++
        j--
      }
    }
    return i

  } else if (srtcompareFn === strDescendingOrderFn) {
    while (i <= j) {
      while (array[i] > pivot) {
        i++
        // console.log('i = ' + i)
      }

      while (array[j] < pivot) {
        j--
        // console.log('j = ' + j)
      }

      if (i <= j) {
        // console.log('swap ' + array[i] + ' with ' + array[j])
        swap(array, i, j)
        i++
        j--
      }
    }
    return i
  }
}

function quick(
  array: unknown[],
  left: number,
  right: number,
  compareFn: CompareFn) {

  let index: number

  if (array.length > 1) {
    index = partition(array, left, right, compareFn)

    if (left < index - 1) {
      quick(array, left, index - 1, compareFn)
    }

    if (index < right) {
      quick(array, index, right, compareFn)
    }
  }
  return array
}

const quickSort = (array: unknown[], compareFn: CompareFn) => {
  return quick(array, 0, array.length - 1, compareFn)
}

const array = [5, 1, 4, 7, 0]
const fn = (a: number, b: number) => b - b
console.log(quickSort(array, fn))

File: Python/src/algorithms/sorting/quick_sort.py
from typing import List


def partition(list: List[int], start: int, end: int) -> int:
    pivot = list[start]
    low = start + 1
    high = end
    while True:
        while low <= high and list[high] >= pivot:
            high = high - 1

        while low <= high and list[low] <= pivot:
            low = low + 1

        if low <= high:
            list[low], list[high] = list[high], list[low]
        else:
            break

    list[start], list[high] = list[high], list[start]
    return high


def quick_sort(list: List[int], start: int, end: int) -> None:
    if start >= end:
        return

    p = partition(list, start, end)
    quick_sort(list, start, p-1)
    quick_sort(list, p+1, end)


list = [29, 99, 27, 41, 66, 28, 44, 78, 87,
        19, 31, 76, 58, 88, 83, 97, 12, 21, 44]
print(list)
quick_sort(list, 0, len(list) - 1)
print(list)
